planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+
+The async daemon however, is a bit lacking (though it does its job ok). The documentation is an old blog post by Jeremy which doesnt say
+that much about it tbh. And you cant shard or put up multiples of it for some sort of HA scenario. It feels a bit scary just running a single service
+with the daemon in production per database. The error handling is a bit funky as well, and if you have an even that crashes a projection somewhere
+in the middle of an event page then actually singling out _that single event_ to have a look at it properly can be very frustrating. I have so far been
+lucky enough to find these weird events in my QA environment so I havent had to do that in production _yet_ but I do not look forward to doing that.
+
+Replaying projections seems to be a thing that you can do, but I find the documentation lacking on this so Im a bit unsure of what actually this does
+which tends to make me prefer spinning up a new projection in parallel and then killing off the old one when it is in sync. It feels safer. But maybe it isnt?
+
+## Unit testing projections
+Projections should be easy to unit test in theory. Apply event A, B, C verify state X, right? But the `IProjection` interface doesnt really make it
+that easy, and the way that they are executed in Marten makes it...harder. In the end I have found it easier to integration test them than it is to 
+unit test them but even while I am not in the "integration tests are evil" camp they are more heavyweight to setup and maintain. And unit testing a
+projection _should_ be a simple thing. But in Marten it is not.
+
+## Running projections against other databases
+One thing that has come up more than once is that we want to run an (async, not inline) projection from an eventstore in database X, but write the projected document
+in database Y. This is ofc impossible today. Main reason for this is that we dont want to give certain apps access to the "important" database but
+rather let them keep their own little sandbox. This could maybe be solved with schemas as well, but then they are still on the same database potentially
+causing load there that we dont want. To be honest, I think this might not necessarily be the easiest problem to solve robustly.
+
+
+# Integrating with an event bus
+
+This is a bit of a special topic but I think that Im not alone in doing this so I think its worth putting my thoughts down on this topic as well. 
+We use MassTransit and RabbitMq and we have probably made _every single mistake that you can do_ publishing events on the bus. 
+
+We started out making the classic mistake of publishing our eventsourced events on the event bus. This was of course a bad move and everyone suffered.
+Eventsourced events != Eventbus events. We ended up building a separate message model where the messages that we post on the bus is enriched with additional
```

Comment:
Been there done that 😉

Reply:
It would be good to have the possibility to transform one event into another (or even map reduce as @jeremydmiller stated somewhere else). Then having events like "RoomReservationInitiated", "RoomReservationDatesUpdated", "RoomReservationConfirmed" we could map it to the "ReservationCreated" event and republish it automatically. In the basic version it might be simmilar to our current transform, but it wouldn't create document but create and store additional event.